// ngx-mic-recorder/esm2020/lib/ngx-mic-recorder.module.mjs
import { NgModule } from '@angular/core';
import { NgxMicRecorderComponent, NgxMicStartStopTemplate, NgxPauseResumeTemplate, } from './ngx-mic-recorder.component';
import { CommonModule } from '@angular/common';
import * as i0 from "@angular/core";
export class NgxMicRecorderModule {
}
NgxMicRecorderModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.7", ngImport: i0, type: NgxMicRecorderModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgxMicRecorderModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.7", ngImport: i0, type: NgxMicRecorderModule, declarations: [NgxMicRecorderComponent,
        NgxMicStartStopTemplate,
        NgxPauseResumeTemplate], imports: [CommonModule], exports: [NgxMicRecorderComponent,
        NgxMicStartStopTemplate,
        NgxPauseResumeTemplate] });
NgxMicRecorderModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.7", ngImport: i0, type: NgxMicRecorderModule, imports: [[
            CommonModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.7", ngImport: i0, type: NgxMicRecorderModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        NgxMicRecorderComponent,
                        NgxMicStartStopTemplate,
                        NgxPauseResumeTemplate,
                    ],
                    imports: [
                        CommonModule
                    ],
                    exports: [
                        NgxMicRecorderComponent,
                        NgxMicStartStopTemplate,
                        NgxPauseResumeTemplate,
                    ]
                }]
        }] });
